README.md
# aprs2influxdb
This program interfaces ham radio APRS-IS servers and saves packet data into an influxdb database. aprs2influxdb handles the connection, parsing, and saving of data into an influxdb database from APRS-IS using line protocol formatted strings. Periodically, a status message is also sent to the APRS-IS server in order to maintain the connection with the APRS-IS server by preventing a timeout.

Supported APRS Packet Formats:
* uncompressed

Non-ASCII characters in APRS packets are ignored!

## Getting started
aprs2influxdb installs using pip can can be installed in editable mode with the source code or from [PyPI](https://pypi.python.org/pypi).

### Prerequisites
You must install and configure an [influxdb](https://www.influxdata.com/) database. Here is their open source [project documentation on GitHub](https://github.com/influxdata/influxdb).

### Installing
It is highly recommended to install `aprs2influxdb` in a virtual environment using `virtualenv`. This helps smooth out installation on Windows and Linux. The executable may not be found and return a "ImportError: 'module' object has no attribute..." indicating that the installed scripts cannot find the installed files. Virtualenv fixes all this. It's not necessary to install in a discrete virtual environment, as long as it is separate from the python installation.

#### PyPI
`pip install aprs2influxdb`

#### Source Code
If you are just installing with source code then navigate to the source directory and run:

`pip install .`

if you are installing in editable mode for development then navigate to the source directory and run:

`pip install -e .`

### Running aprs2influxdb
The program defaults use standard influxdb login information as well as example APRS-IS login information. If you properly installed influxdb you will need to specify your own database information. Additionally, you will need an amateur radio license with which you may login to APRS-IS with your callsign.

#### Command Line Options

* `--help` show this help message and exit
* `--dbhost DBHOST` set influxdb host (default = localhost)
* `--dbport DBPORT` set influxdb port (default = 8086)
* `--dbuser DBUSER` set influxdb user (default = root)
* `--dbpassword DBPASSWORD` set influxdb password (default = root)
* `--dbname DBNAME` set influxdb database name (default = mydb)
* `--callsign CALLSIGN` set APRS-IS login callsign (default = nocall)
* `--port PORT` set APRS-IS port (default = 10152)
* `--interval INTERVAL` set APRS-IS heartbeat interval in minutes (default = 15)
* `--debug` Set logging level to DEBUG (default = False)

#### Example
Starting aprs2influxdb assuming an influxdb server is running and has a "mydb" database configured is simple. Please note that APRS-IS ignores logins from "nocall" so you will connect but likely see nothing if you do not specify your amateur radio callsign.

`aprs2influxdb --dbuser influxuser --dbpassword password123 --dbname mydb --callsign nocall`

The above command uses default values for the options not specified. APRS-IS port 10152 is the full stream while other ports exist this is the most useful. aprslib defaults to `rotate.aprs.net` to pick an APRS core server. Please see [APRS-IS Servers](http://www.aprs-is.net/aprsservers.aspx) for more information.

To exit `aprs2influxdb` use `cntl+c` on git bash for Windows and `cntl+z` followed by `kill <pid>` for the PID used by `aprs2influxdb` on Linux.

## Running the tests

Unit testing will be implemented in a future pull request

## Deployment
This has been tested on a Debian 9 (Stretch) server as well as locally with Windows 7 during development.

## Authors
* **Bryce Salmi** - *Initial work* - [KB1LQC](https://github.com/kb1lqc)

See also the list of [contributors](https://github.com/FaradayRF/aprs2influxdb/contributors) who participated in this project.

## Acknowledgments

* [@PhirePhly](https://github.com/PhirePhly) for answering my APRS questions!
* [@hessu](https://github.com/hessu) for also answering my APRS and aprsc questions as well as providing the awesome [aprs.fi](https://www.aprs.fi) website.


CHANGES.txt
v0.0.1, 9-27-2017 -- Initial Release.


clerk.py
import aprslib
import configparser
import influxdb
from influxdb import InfluxDBClient
import Geohash
import time
import logging

# Globals
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("DrWatson")

def jsonToLineProtocol(jsonData):
    # Converts aprslib JSON to influxdb line protocol
    # Schema
    # measurement = packet
    # tag = from
    # tag = to
    # tag = symbolTable
    # tag = symbol
    # tag = format
    # tag = comment
    # field = latitude
    # field = longitude
    # field = posAmbiguity
    # field = altitude
    # field = speed
    # field = sequenceNumber
    # field = analog1
    # field = analog2
    # field = analog3
    # field = analog4
    # field = analog5
    # field = digital

    if jsonData["format"] == "uncompressed":
        # initialize variables
        tags = []
        fields = []

        # Set measurement to "packet"
        measurement = "packet"

        try:
            tags.append("from={0}".format(jsonData.get("from")))
            tags.append("to={0}".format(jsonData.get("to")))
	    #symbolTable = jsonData.get("symbol_table")
	    #tags.append("symbolTable=\"{0}\"".format(symbolTable.encode("ascii", errors="replace")))
            #symbol = jsonData.get("symbol")
            #tags.append("symbol=\"{0}\"".format(symbol.encode("ascii", errors="replace")))
            tags.append("format={0}".format(jsonData.get("format")))

        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        try:
	    rawComment = jsonData.get("comment")
	    comment = rawComment.encode("ascii", errors="ignore")
	    #tags.append("comment=\"{0}\"".format(comment))

        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        except UnicodeError as e:
            logger.error(e)
            logger.error(jsonData)

        tagStr = ",".join(tags)

        try:
            fields.append("latitude={0}".format(jsonData.get("latitude", 0)))
            fields.append("longitude={0}".format(jsonData.get("longitude", 0)))
            fields.append("posAmbiguity={0}".format(jsonData.get("posambiguity", 0)))
            fields.append("altitude={0}".format(jsonData.get("altitude",0)))
            fields.append("speed={0}".format(jsonData.get("speed", 0)))
        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        try:
            if jsonData["telemetry"]["seq"]:
                fields.append("sequenceNumber={0}".format(jsonData["telemetry"]["seq"]))
                fields.append("analog1={0}".format(jsonData["telemetry"]["vals"][0]))
                fields.append("analog2={0}".format(jsonData["telemetry"]["vals"][1]))
                fields.append("analog3={0}".format(jsonData["telemetry"]["vals"][2]))
                fields.append("analog4={0}".format(jsonData["telemetry"]["vals"][3]))
                fields.append("analog5={0}".format(jsonData["telemetry"]["vals"][4]))
                fields.append("digital={0}".format(jsonData["telemetry"]["bits"]))

        except KeyError as e:
            # Expect many KeyErrors for stations not sending telemetry
            logger.debug(e)
            logger.debug(jsonData)

        fieldsStr = ",".join(fields)

	return measurement + "," + tagStr + " " + fieldsStr


def callback(packet):
    packet = aprslib.parse(packet)
    # Open a new connection every time, probably slow
    influxConn = connectInfluxDB()
    line = jsonToLineProtocol(packet)

    if line:
        logger.debug(line)
    	try:
            influxConn.write_points([line], protocol='line')

        except StandardError as e:
            logger.error(e)
            #logger.error(jsonData)

        except influxdb.exceptions.InfluxDBClientError as e:
            logger.error(e)
            #logger.error(jsonData)


def connectInfluxDB():

    config = configparser.ConfigParser()
    config.read('clerk.ini')

    host = config['influx']['host']
    port = config['influx']['port']
    user = config['influx']['user']
    password = config['influx']['password']
    dbname = config['influx']['dbname']
    dbuser = config['influx']['dbuser']
    dbuser_password = config['influx']['dbuserpassword']

    return InfluxDBClient(host, port, user, password, dbname)

def main():
    # Start logger

    # Open APRS-IS connection
    AIS = aprslib.IS("KB1LQC")
    AIS.connect()

    # Obtain raw APRS-IS packets and sent to callback when received
    AIS.consumer(callback, raw=True)

if __name__ == "__main__":
    main()


aprs2influxdb.py
import aprslib
import ConfigParser
import influxdb
from influxdb import InfluxDBClient
import logging
import argparse
import os

# Globals
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("aprs2influxdb")

# Command line input
parser = argparse.ArgumentParser(description='Connects to APRS-IS and saves stream to local InfluxDB')
parser.add_argument('--init-config', dest='init', action='store_true', help='Initialize configuration file')
parser.add_argument('--host', help='Set InfluxDB host')
parser.add_argument('--port', help='Set InfluxDB port')
parser.add_argument('--user', help='Set InfluxDB user')
parser.add_argument('--password', help='Set InfluxDB password')
parser.add_argument('--dbname', help='Set InfluxDB database name')
parser.add_argument('--dbuser', help='Set InfluxDB user')
parser.add_argument('--dbuserpassword', help='Set InfluxDB user password')

# Parse the arguments
args = parser.parse_args()

logger.info("woot")

def getConfig():
        """
        Get configuration file
        """
        # Known paths where loggingConfig.ini can exist
        relpath1 = os.path.join('etc', 'aprs2influxdb')

        # Check all directories until first instance of loggingConfig.ini
        for location in os.curdir, relpath1:
            try:
                config = ConfigParser.RawConfigParser()
                result = config.read(os.path.join(location, "config.ini"))
            except ConfigParser.NoSectionError:
                pass

            if result:
                break

        return config

def jsonToLineProtocol(jsonData):
    # Converts aprslib JSON to influxdb line protocol
    # Schema
    # measurement = packet
    # tag = from
    # tag = to
    # tag = symbolTable
    # tag = symbol
    # tag = format
    # tag = comment
    # field = latitude
    # field = longitude
    # field = posAmbiguity
    # field = altitude
    # field = speed
    # field = sequenceNumber
    # field = analog1
    # field = analog2
    # field = analog3
    # field = analog4
    # field = analog5
    # field = digital

    if jsonData["format"] == "uncompressed":
        # initialize variables
        tags = []
        fields = []

        # Set measurement to "packet"
        measurement = "packet"

        try:
            tags.append("from={0}".format(jsonData.get("from")))
            tags.append("to={0}".format(jsonData.get("to")))
            #symbolTable = jsonData.get("symbol_table")
            #tags.append("symbolTable=\"{0}\"".format(symbolTable.encode("ascii", errors="replace")))
            #symbol = jsonData.get("symbol")
            #tags.append("symbol=\"{0}\"".format(symbol.encode("ascii", errors="replace")))
            tags.append("format={0}".format(jsonData.get("format")))

        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        # try:
        #     rawComment = jsonData.get("comment")
        #     comment = rawComment.encode("ascii", errors="ignore")
        #     tags.append("comment=\"{0}\"".format(comment))
        #
        # except KeyError as e:
        #     logger.error(e)
        #     logger.error(jsonData)
        #
        # except UnicodeError as e:
        #     logger.error(e)
        #     logger.error(jsonData)

        tagStr = ",".join(tags)

        try:
            fields.append("latitude={0}".format(jsonData.get("latitude", 0)))
            fields.append("longitude={0}".format(jsonData.get("longitude", 0)))
            fields.append("posAmbiguity={0}".format(jsonData.get("posambiguity", 0)))
            fields.append("altitude={0}".format(jsonData.get("altitude", 0)))
            fields.append("speed={0}".format(jsonData.get("speed", 0)))
        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        try:
            if jsonData["telemetry"]["seq"]:
                fields.append("sequenceNumber={0}".format(jsonData["telemetry"]["seq"]))
                fields.append("analog1={0}".format(jsonData["telemetry"]["vals"][0]))
                fields.append("analog2={0}".format(jsonData["telemetry"]["vals"][1]))
                fields.append("analog3={0}".format(jsonData["telemetry"]["vals"][2]))
                fields.append("analog4={0}".format(jsonData["telemetry"]["vals"][3]))
                fields.append("analog5={0}".format(jsonData["telemetry"]["vals"][4]))
                fields.append("digital={0}".format(jsonData["telemetry"]["bits"]))

        except KeyError as e:
            # Expect many KeyErrors for stations not sending telemetry
            logger.debug(e)
            logger.debug(jsonData)

        fieldsStr = ",".join(fields)

        return measurement + "," + tagStr + " " + fieldsStr


def callback(packet):
    packet = aprslib.parse(packet)
    # Open a new connection every time, probably slow
    influxConn = connectInfluxDB()
    line = jsonToLineProtocol(packet)

    if line:
        logger.debug(line)
        try:
            influxConn.write_points([line], protocol='line')

        except StandardError as e:
            logger.error(e)
            #logger.error(jsonData)

        except influxdb.exceptions.InfluxDBClientError as e:
            logger.error(e)
            #logger.error(jsonData)


def connectInfluxDB():
    configFile = getConfig()
    host = configFile.get('influx','host')
    port = configFile.get('influx','port')
    user = configFile.get('influx','user')
    password = configFile.get('influx','password')
    dbname = configFile.get('influx','dbname')
    #dbuser = config['influx']['dbuser']
    #dbuser_password = config['influx']['dbuserpassword']

    return InfluxDBClient(host, port, user, password, dbname)


def main():
    # Start logger


    # Open APRS-IS connection
    AIS = aprslib.IS("KB1LQC")
    AIS.connect()

    # Obtain raw APRS-IS packets and sent to callback when received
    AIS.consumer(callback, raw=True)


if __name__ == "__main__":
    main()


aprs2influxdb/aprs2influxdbscript.py
import aprslib
import influxdb
from influxdb import InfluxDBClient
import logging
import argparse
import os
import sys
import threading
import time

# Globals
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("aprs2influxdb")

# Command line input
parser = argparse.ArgumentParser(description='Connects to APRS-IS and saves stream to local InfluxDB')
parser.add_argument('--dbhost', help='Set InfluxDB host', default="localhost")
parser.add_argument('--dbport', help='Set InfluxDB port', default="8086")
parser.add_argument('--dbuser', help='Set InfluxDB user', default="root")
parser.add_argument('--dbpassword', help='Set InfluxDB password', default="root")
parser.add_argument('--dbname', help='Set InfluxDB database name', default="mydb")
parser.add_argument('--callsign', help='Set APRS-IS login callsign', default="nocall")
parser.add_argument('--port', help='Set APRS-IS port', default="10152")
parser.add_argument('--interval', help='Set APRS-IS heartbeat interval in minutes', default="15")

# Parse the arguments
args = parser.parse_args()

def jsonToLineProtocol(jsonData):
    """Converts JSON APRS-IS packet to influxdb line protocol

    Takes in a JSON packet from aprslib (raw=false) and parses it into an
    influxdb line protocol compliant string to insert into database. Returns
    a valid line protocol string ready to be inserted into the database.

    keyword arguments:
    jsonData -- aprslib parsed JSON packet
    """
    # Converts aprslib JSON to influxdb line protocol
    # Schema
    # measurement = packet
    # tag = from
    # tag = to
    # tag = symbolTable
    # tag = symbol
    # tag = format
    # tag = comment
    # field = latitude
    # field = longitude
    # field = posAmbiguity
    # field = altitude
    # field = speed
    # field = sequenceNumber
    # field = analog1
    # field = analog2
    # field = analog3
    # field = analog4
    # field = analog5
    # field = digital

    # Parse uncompressed format packets
    if jsonData["format"] == "uncompressed":
        # initialize variables
        tags = []
        fields = []

        # Set measurement to "packet"
        measurement = "packet"

        try:
            tags.append("from={0}".format(jsonData.get("from")))
            tags.append("to={0}".format(jsonData.get("to")))
            tags.append("format={0}".format(jsonData.get("format")))

        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        tagStr = ",".join(tags)

        try:
            fields.append("latitude={0}".format(jsonData.get("latitude", 0)))
            fields.append("longitude={0}".format(jsonData.get("longitude", 0)))
            fields.append("posAmbiguity={0}".format(jsonData.get("posambiguity", 0)))
            fields.append("altitude={0}".format(jsonData.get("altitude", 0)))
            fields.append("speed={0}".format(jsonData.get("speed", 0)))
        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        try:
            if jsonData["telemetry"]["seq"]:
                fields.append("sequenceNumber={0}".format(jsonData["telemetry"]["seq"]))
                fields.append("analog1={0}".format(jsonData["telemetry"]["vals"][0]))
                fields.append("analog2={0}".format(jsonData["telemetry"]["vals"][1]))
                fields.append("analog3={0}".format(jsonData["telemetry"]["vals"][2]))
                fields.append("analog4={0}".format(jsonData["telemetry"]["vals"][3]))
                fields.append("analog5={0}".format(jsonData["telemetry"]["vals"][4]))
                fields.append("digital={0}".format(jsonData["telemetry"]["bits"]))

        except KeyError as e:
            # Expect many KeyErrors for stations not sending telemetry
            logger.debug(e)
            logger.debug(jsonData)

        try:
            comment = jsonData.get("comment").encode('ascii', 'ignore')

            if comment:
                logger.debug(comment)
                fields.append("comment=\"{0}\"".format(comment.replace("\"", "")))

        except UnicodeError as e:
            logger.error(e)

        except TypeError as e:
            logger.error(e)

        fieldsStr = ",".join(fields)

        return measurement + "," + tagStr + " " + fieldsStr


def callback(packet):
    """aprslib callback for every packet received from APRS-IS connection

    keyword arguments:
    packet -- APRS-IS packet from aprslib connection
    """
    logger.debug(packet)

    # Open a new connection every time, probably SLOWWWW
    influxConn = connectInfluxDB()
    line = jsonToLineProtocol(packet)

    if line:
        logger.debug(line)
        try:
            influxConn.write_points([line], protocol='line')

        except StandardError as e:
            logger.error(e)
            logger.error(packet)

        except influxdb.exceptions.InfluxDBClientError as e:
            logger.error(e)
            logger.error(packet)


def connectInfluxDB():
    """Connect to influxdb database with configuration values"""

    return InfluxDBClient(  args.dbhost,
                            args.dbport,
                            args.dbuser,
                            args.dbpassword,
                            args.dbname)


def consumer(conn):
    """Start consumer function for thread

    keyword arguments:
    conn -- APRS-IS connection from aprslib
    """
    logger.debug("starting consumer thread")
    # Obtain raw APRS-IS packets and sent to callback when received
    conn.consumer(callback, immortal=True, raw=False)


def heartbeat(conn, callsign, interval):
    """Send out an APRS status message to keep connection alive

    keyword arguments:
    conn -- APRS-IS connction from aprslib
    callsign -- Callsign of status message
    interval -- Minutes betwee status messages
    """
    logger.debug("Starting heartbeat thread")
    while True:
        # Create timestamp
        timestamp = int(time.time())

        # Create APRS status message
        status = "{0}>APRS,TCPIP*:>aprs2influxdb heartbeat {1}"
        conn.sendall(status.format(callsign, timestamp))
        logger.debug("Sent heartbeat")

        # Sleep for specified time
        time.sleep(float(interval) * 60)  # Sent every interval minutes


def main():
    """Main function of aprs2influxdb

    Reads in configuration values and starts connection to APRS-IS with aprslib.
    Then two threads are started, one to monitor for APRS-IS packets and
    another to periodically send status packets to APRS-IS in order to keep
    the connection alive.
    """

    # Start login for APRS-IS
    logger.info("Logging into APRS-IS as {0} on port {1}".format(args.callsign, args.port))
    if args.callsign == "nocall":
        logger.warning("APRS-IS ignores the callsign \"nocall\"!")

    # Open APRS-IS connection
    passcode = aprslib.passcode(args.callsign)
    AIS = aprslib.IS(   args.callsign,
                        passwd=passcode,
                        port=args.port)
    try:
        AIS.connect()

    except aprslib.exceptions.LoginError as e:
        logger.error(e)
        logger.info("APRS Login Callsign: {0} Port: {1}".format(args.callsign, args.port))
        sys.exit(1)

    except aprslib.exceptions.ConnectionError as e:
        logger.error(e)
        sys.exit(1)

    # Create heartbeat
    t1 = threading.Thread(target=heartbeat, args=(AIS, args.callsign, args.interval))

    # Create consumer
    t2 = threading.Thread(target=consumer, args=(AIS,))

    # Start threads
    t1.start()
    t2.start()


if __name__ == "__main__":
    main()


aprs2influxdb/main.py
import aprslib
import ConfigParser
import influxdb
from influxdb import InfluxDBClient
import logging
import argparse
import os
import sys
import threading
import time

# Globals
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("aprs2influxdb")

# Command line input
parser = argparse.ArgumentParser(description='Connects to APRS-IS and saves stream to local InfluxDB')
parser.add_argument('--dbhost', help='Set InfluxDB host', default="localhost")
parser.add_argument('--dbport', help='Set InfluxDB port', default="8086")
parser.add_argument('--dbuser', help='Set InfluxDB user', default="root")
parser.add_argument('--dbpassword', help='Set InfluxDB password', default="root")
parser.add_argument('--dbname', help='Set InfluxDB database name', default="mydb")
parser.add_argument('--callsign', help='Set APRS-IS login callsign', default="nocall")
parser.add_argument('--port', help='Set APRS-IS port', default="10152")
parser.add_argument('--interval', help='Set APRS-IS heartbeat interval in minutes', default="15")

# Parse the arguments
args = parser.parse_args()


def editConfig(config, args):
    """Edits the configuration file based on command line arguments

    keyword arguments:
    config -- configuration file descriptor
    args -- argparse arguments
    """
    #Use command line values to change configuration
    if args.dbhost:
        config[0].set('influx', 'dbhost', args.dbhost)
    if args.dbport:
        config[0].set('influx', 'dbport', args.dbport)
    if args.dbuser:
        config[0].set('influx', 'dbuser', args.dbuser)
    if args.dbpassword:
        config[0].set('influx', 'dbpassword', args.dbpassword)
    if args.dbname:
        config[0].set('influx', 'dbname', args.dbname)
    if args.callsign:
        config[0].set('aprsis', 'callsign', args.callsign)
    if args.port:
        config[0].set('aprsis', 'port', args.port)
    if args.interval:
        config[0].set('aprsis', 'interval', args.interval)

    with open(config[1], 'wb') as configfile:
        config[0].write(configfile)


def getConfig():
        """Open configuration file and return a list of config database

        Searches the current directory as well as Python specific install
        location such as /etc/<installPath> or C:/python27/<installPath>.
        Returns a list of config file desciptor as well as the path it was found
        at so other operations can use it to open the file.
        """
        # Known paths where loggingConfig.ini can exist
        installPath = os.path.join(sys.prefix, "etc", "aprs2influxdb", "config.ini")
        localPath = os.path.join(os.curdir, "config.ini")

        # Check all directories until first instance of loggingConfig.ini
        for location in localPath, installPath:
            try:
                config = ConfigParser.RawConfigParser()
                result = config.read(location)
            except ConfigParser.NoSectionError:
                pass

            if result:
                break

        return [config, location]


def jsonToLineProtocol(jsonData):
    """Converts JSON APRS-IS packet to influxdb line protocol

    Takes in a JSON packet from aprslib (raw=false) and parses it into an
    influxdb line protocol compliant string to insert into database. Returns
    a valid line protocol string ready to be inserted into the database.

    keyword arguments:
    jsonData -- aprslib parsed JSON packet
    """
    # Converts aprslib JSON to influxdb line protocol
    # Schema
    # measurement = packet
    # tag = from
    # tag = to
    # tag = symbolTable
    # tag = symbol
    # tag = format
    # tag = comment
    # field = latitude
    # field = longitude
    # field = posAmbiguity
    # field = altitude
    # field = speed
    # field = sequenceNumber
    # field = analog1
    # field = analog2
    # field = analog3
    # field = analog4
    # field = analog5
    # field = digital

    # Parse uncompressed format packets
    if jsonData["format"] == "uncompressed":
        # initialize variables
        tags = []
        fields = []

        # Set measurement to "packet"
        measurement = "packet"

        try:
            tags.append("from={0}".format(jsonData.get("from")))
            tags.append("to={0}".format(jsonData.get("to")))
            tags.append("format={0}".format(jsonData.get("format")))

        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        tagStr = ",".join(tags)

        try:
            fields.append("latitude={0}".format(jsonData.get("latitude", 0)))
            fields.append("longitude={0}".format(jsonData.get("longitude", 0)))
            fields.append("posAmbiguity={0}".format(jsonData.get("posambiguity", 0)))
            fields.append("altitude={0}".format(jsonData.get("altitude", 0)))
            fields.append("speed={0}".format(jsonData.get("speed", 0)))
        except KeyError as e:
            logger.error(e)
            logger.error(jsonData)

        try:
            if jsonData["telemetry"]["seq"]:
                fields.append("sequenceNumber={0}".format(jsonData["telemetry"]["seq"]))
                fields.append("analog1={0}".format(jsonData["telemetry"]["vals"][0]))
                fields.append("analog2={0}".format(jsonData["telemetry"]["vals"][1]))
                fields.append("analog3={0}".format(jsonData["telemetry"]["vals"][2]))
                fields.append("analog4={0}".format(jsonData["telemetry"]["vals"][3]))
                fields.append("analog5={0}".format(jsonData["telemetry"]["vals"][4]))
                fields.append("digital={0}".format(jsonData["telemetry"]["bits"]))

        except KeyError as e:
            # Expect many KeyErrors for stations not sending telemetry
            logger.debug(e)
            logger.debug(jsonData)

        try:
            comment = jsonData.get("comment").encode('ascii', 'ignore')

            if comment:
                logger.debug(comment)
                fields.append("comment=\"{0}\"".format(comment.replace("\"", "")))

        except UnicodeError as e:
            logger.error(e)

        except TypeError as e:
            logger.error(e)

        fieldsStr = ",".join(fields)

        return measurement + "," + tagStr + " " + fieldsStr


def callback(packet):
    """aprslib callback for every packet received from APRS-IS connection

    keyword arguments:
    packet -- APRS-IS packet from aprslib connection
    """
    logger.debug(packet)

    # Open a new connection every time, probably SLOWWWW
    influxConn = connectInfluxDB()
    line = jsonToLineProtocol(packet)

    if line:
        logger.debug(line)
        try:
            influxConn.write_points([line], protocol='line')

        except StandardError as e:
            logger.error(e)
            logger.error(packet)

        except influxdb.exceptions.InfluxDBClientError as e:
            logger.error(e)
            logger.error(packet)


def connectInfluxDB():
    """Connect to influxdb database with configuration values"""
    config = getConfig()
    configFile = config[0]
    host = configFile.get('influx', 'dbhost')
    port = configFile.get('influx', 'dbport')
    user = configFile.get('influx', 'dbuser')
    password = configFile.get('influx', 'dbpassword')
    dbname = configFile.get('influx', 'dbname')

    return InfluxDBClient(host, port, user, password, dbname)


def consumer(conn):
    """Start consumer function for thread

    keyword arguments:
    conn -- APRS-IS connection from aprslib
    """
    logger.debug("starting consumer thread")
    # Obtain raw APRS-IS packets and sent to callback when received
    conn.consumer(callback, immortal=True, raw=False)


def heartbeat(conn, callsign, interval):
    """Send out an APRS status message to keep connection alive

    keyword arguments:
    conn -- APRS-IS connction from aprslib
    callsign -- Callsign of status message
    interval -- Minutes betwee status messages
    """
    logger.debug("Starting heartbeat thread")
    while True:
        # Create timestamp
        timestamp = int(time.time())

        # Create APRS status message
        status = "{0}>APRS,TCPIP*:>aprs2influxdb heartbeat {1}"
        conn.sendall(status.format(callsign, timestamp))
        logger.debug("Sent heartbeat")

        # Sleep for specified time
        time.sleep(interval * 60)  # Sent every interval minutes


def main():
    """Main function of aprs2influxdb

    Reads in configuration values and starts connection to APRS-IS with aprslib.
    Then two threads are started, one to monitor for APRS-IS packets and
    another to periodically send status packets to APRS-IS in order to keep
    the connection alive.
    """
    # Open up configuration file
    config = getConfig()

    # Edit configuration with user input
    editConfig(config, args)

    # Obtain APRS-IS configuration
    aprsCallsign = config[0].get('aprsis', 'callsign').upper()
    aprsPort = config[0].get('aprsis', 'port')
    passcode = aprslib.passcode(aprsCallsign)
    aprsInterval = config[0].getint('aprsis', 'interval')

    # Open APRS-IS connection
    AIS = aprslib.IS(aprsCallsign, passwd=passcode, port=aprsPort)
    try:
        AIS.connect()

    except aprslib.exceptions.LoginError as e:
        logger.error(e)
        logger.info("APRS Login Callsign: {0} Port: {1}".format(aprsCallsign, aprsPort))
        sys.exit(1)

    except aprslib.exceptions.ConnectionError as e:
        logger.error(e)
        sys.exit(1)

    # Create heartbeat
    t1 = threading.Thread(target=heartbeat, args=(AIS, aprsCallsign, aprsInterval))

    # Create consumer
    t2 = threading.Thread(target=consumer, args=(AIS,))

    # Start threads
    t1.start()
    t2.start()


if __name__ == "__main__":
    main()
